File: app/src/main/java/com/timotiusoktorio/pencake/ui/menu/MenuFragmentViewModel.java
package com.timotiusoktorio.pencake.ui.menu;

import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;

import com.timotiusoktorio.pencake.data.model.Category;
import com.timotiusoktorio.pencake.data.model.State;
import com.timotiusoktorio.pencake.data.source.DataManager;

import java.util.List;

class MenuFragmentViewModel extends ViewModel {

    final MutableLiveData<List<Category>> categoriesLiveData = new MutableLiveData<>();
    final MutableLiveData<State> stateLiveData = new MutableLiveData<>();

    private final DataManager dataManager;

    MenuFragmentViewModel(DataManager dataManager) {
        this.dataManager = dataManager;
        loadCategories();
    }

    private void loadCategories() {
        stateLiveData.setValue(State.LOADING);
        dataManager.fetchCategories(new DataManager.Callback<Category>() {
            @Override
            public void onSuccess(List<Category> data) {
                categoriesLiveData.setValue(data);
                stateLiveData.setValue(State.SUCCESS);
            }

            @Override
            public void onError(String errorMsg) {
                stateLiveData.setValue(State.ERROR);
            }
        });
    }
}

File: app/src/main/java/com/timotiusoktorio/pencake/ui/menu/MenuFragmentViewModel.kt
package com.timotiusoktorio.pencake.ui.menu

import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.ViewModel
import com.timotiusoktorio.pencake.data.model.Category
import com.timotiusoktorio.pencake.data.model.State
import com.timotiusoktorio.pencake.data.source.DataManager

class MenuFragmentViewModel(private val dataManager: DataManager) : ViewModel() {

    val stateLiveData = MutableLiveData<State>()
    val categoriesLiveData = MutableLiveData<List<Category>>()

    init {
        loadCategories()
    }

    private fun loadCategories() {
        stateLiveData.value = State.LOADING
        dataManager.fetchCategories(object : DataManager.Callback<Category> {
            override fun onSuccess(data: List<Category>) {
                categoriesLiveData.value = data
                stateLiveData.value = State.SUCCESS
            }

            override fun onError(errorMsg: String) {
                stateLiveData.value = State.ERROR
            }
        })
    }
}